python_etl/churn.py

```diff
@@ -0,0 +1,654 @@
+    attribution_fields = ["source", "medium", "campaign", "content"]
+    if datum.attribution:
+        for field in attribution_fields:
+            out[field] = datum.attribution[field]
+
+    # bug 1323598
+    if datum.distribution_id:
+        out['distribution_id'] = datum.distribution_id
+    if datum.default_search_engine:
+        out['default_search_engine'] = datum.default_search_engine
+    if datum.locale:
+        out['locale'] = datum.locale
+
+    deviceCount = 0
+    if datum.sync_count_desktop is not None:
+        deviceCount += datum.sync_count_desktop
+    if datum.sync_count_mobile is not None:
+        deviceCount += datum.sync_count_mobile
+
+    if deviceCount > 1:
+        out["sync_usage"] = "multiple"
+    elif deviceCount == 1:
+        out["sync_usage"] = "single"
+    elif datum.sync_configured is not None:
+        if datum.sync_configured:
+            out["sync_usage"] = "single"
+        else:
+            out["sync_usage"] = "no"
+    # Else we don't set sync_usage at all, and use a default value later.
+
+    out["current_version"] = datum.version
+
+    # The usage time is in seconds, but we really need hours.  Because
+    # we filter out broken subsession_lengths, we could end up with
+    # clients with no usage hours.
+    out["usage_hours"] = ((datum.usage_seconds / SECONDS_IN_HOUR)
+                          if datum.usage_seconds is not None
+                          else 0.0)
+    out["squared_usage_hours"] = out["usage_hours"] ** 2
+
+    out["total_uri_count"] = datum.total_uri_count_per_client
+    out["unique_domains_count"] = datum.average_unique_domains_count_per_client
+
+    # Incoming subsession_start_date looks like "2016-02-22T00:00:00.0-04:00"
+    client_date = None
+    if datum.subsession_start_date is not None:
+        try:
+            client_date = datetime.strptime(
+                datum.subsession_start_date[0:10], "%Y-%m-%d")
+        except ValueError:
+            # Bogus format
+            pass
+        except TypeError:
+            # String contains null bytes or other weirdness. Example:
+            # TypeError: must be string without null bytes, not unicode
+            pass
+    if client_date is None:
+        # Fall back to submission date
+        client_date = datetime.strptime(datum.submission_date_s3, "%Y%m%d")
+    out["current_week"] = get_week_num(pcd, client_date)
+    out["is_active"] = "yes"
+    if client_date is not None:
+        try:
+            if datetime.strftime(client_date, "%Y%m%d") < week_start:
+                out["is_active"] = "no"
+        except ValueError:
+            pass
+    out["good"] = True
+    return out
+
+
+# Build the "effective version" cache:
+d2v = make_d2v(get_release_info())
+
+
+def fmt(d, date_format="%Y%m%d"):
+    return datetime.strftime(d, date_format)
+
+
+# ### Compute the aggregates
+#
+# Run the aggregation code, detecting files that are missing.
+#
+# The fields we want in the output are:
```

Comment:
Please move these field descriptions down next to the `churn_schema` declaration.